realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -78,6 +86,88 @@ static RealmCacheType valueOf(Class<? extends BaseRealm> clazz) {
         }
     }
 
+    private static class CreateRealmRunnable<T extends BaseRealm> implements Runnable {
+        private RealmConfiguration configuration;
+        private BaseRealm.InstanceCallback<T> callback;
+        private Class<T> realmClass;
+        private CountDownLatch createdInForegroundLatch = new CountDownLatch(1);
+        private RealmNotifier notifier;
+        // The Future this runnable belongs to.
+        private Future future;
+
+        CreateRealmRunnable(RealmNotifier notifier, RealmConfiguration configuration,
+                            BaseRealm.InstanceCallback<T> callback, Class<T> realmClass) {
+            this.configuration = configuration;
+            this.realmClass = realmClass;
+            this.callback = callback;
+            this.notifier = notifier;
+        }
+
+        public void setFuture(Future future) {
+            this.future = future;
+        }
+
+        @Override
+        public void run() {
+            T instance = null;
+            try {
+                instance = createRealmOrGetFromCache(configuration, realmClass);
+                boolean results = notifier.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        // If the RealmAsyncTask.cancel() is called before, we just return without creating the Realm
+                        // instance on the caller thread.
+                        // Thread.isInterrupted() cannot be used for checking here since CountDownLatch.await() will
+                        // will clear interrupted status.
+                        // Using the future to check which this runnable belongs to is to ensure if it is canceled from
+                        // the caller thread before, the callback will never be delivered.
+                        if (future == null || future.isCancelled()) {
+                            return;
+                        }
+                        T instanceToReturn = null;
+                        try {
+                            instanceToReturn = createRealmOrGetFromCache(configuration, realmClass);
+                        } catch (Throwable e) {
+                            callback.onError(e);
+                        } finally {
+                            createdInForegroundLatch.countDown();
```

Comment:
Shouldn't we call `createdInForegroundLatch.countDown();` before `callback.onError(e);`?

```
T instanceToReturn = null;
Throwable error = null;
try {
    instanceToReturn = ...;
} catch (Throwable e) {
    error = e;
} finally {
    createdInForegroundLatch.countDown();
}

if (instanceToReturn != null) {
    callback.onSuccess(instanceToReturn);
} else {
    callback.onError(error);
}
```